src/librustc_mir/hair/pattern/check_match.rs

```diff
@@ -229,7 +229,7 @@ impl<'a, 'tcx> MatchVisitor<'a, 'tcx> {
                 let scrutinee_is_uninhabited = if self.tcx.features().exhaustive_patterns {
                     self.tcx.is_ty_uninhabited_from(module, pat_ty)
                 } else {
-                    self.conservative_is_uninhabited(pat_ty)
+                    pat_ty.conservative_is_uninhabited(self.tcx)
```

Comment:
Again, I think privacy ought to be considered here. It seems to me that this `conservative_is_uninhabited` function has been misnamed, at least as presently implemented. It's conservative along one dimension, but not in terms of privacy etc. It feels a bit misleading. Perhaps `conservative_is_privately_uninhabited` or something? In any case, the `is_ty_uninhabited_from` call above was *precisely* intended to capture this "privacy angle" here. Ironically, with the code as we have it here, enabling `exhaustive_patterns` makes us start to consider privacy again -- maybe compare the behavior of the test `src/test/ui/match/match-privately-empty.rs`, for example, when we remove the `exhaustive_patterns` feature gate?

Reply:
Do you think the previous version of `conservative_is_uninhabited` was just wrong? In accordance with other places `exhaustive_patterns` is checked for, a type is simply considered always inhabited if `exhaustive_patterns` is not considered. So this should probably just be:
```rust
pat_ty.is_never()
```

Reply:
I sort of forget the details here, but I think that the existing behavior (no feature gate) is something like:

- If you are matching **with no arms** on something "privately uninhabited", we error ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2015&gist=0d45cacaf51a2f164b164614943bfca9)
- If you are matching **with no arms** on something "publicly uninhabited", it is ok ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2015&gist=193737f9d04b6f117f69ff5c60fcc828))

but I think the reasoning about private/public was sort of implicit in the definition. It had to be either `!` (clearly uninhabited) or an empty enum -- in order to have an expression of empty enum type at this point, that empty enum must be "public enough" to be seen by this module though.